### PR TITLE
Add some more structure to the Ruby files

### PR DESCRIPTION
**Pull request by iHiD**
This does the following:
- Adds some extra structure by moving the stack, architecture, etc files into lib and adding a base file.
- Adds use of `private` and `attr_accessor` to make my life easier when writing tests for this.
- Also adds Rubocop and Prettier and the git hook in line with the other v3 repos.
- Fixes the non-building Dockerfile ([Enjoy!!](https://github.com/moby/moby/issues/37965#issuecomment-426853382))
- Adds the exercism-config gem to the Gemfile. Bumping it here becomes the official way to bump it in the Dockerfile.

I think we should probably think about adding Groups to the Gemfile along with the other non-rails things.

@joshgoebel Any thoughts or can I merge this and we keep moving forward?

### CODE REVIEW

File: .github/workflows/dockerhub.yml [hunk 1]
```diff
@@ -12,6 +12,12 @@ jobs:
       - name: Checkout code
         uses: actions/checkout@v2
 
+      - name: Set exercism config version
+        run: echo ::set-env name=EXERCISM_CONFIG_VERSION::$(grep 'exercism-config (' Gemfile.lock | grep -o '[0-9.]\+')
```
Comment (joshgoebel) [inline .github/workflows/dockerhub.yml:16]: Hahahaha, great.

File: setup/src/shell [hunk 2]
```diff
@@ -1,4 +1,5 @@
 #!/bin/bash
+# echo
```
Comment (joshgoebel) [inline setup/src/shell:2]: This could probably be removed since it's in a file now.

Comment (iHiD) [reply setup/src/shell:2]: Let's do a second PR for this :)

**Comment by iHiD** [comment]
Sweet. Thanks! :)

**Pull request merged**